content/blog/2019-09-25-vril-vector-bril/index.md

```diff
@@ -0,0 +1,137 @@
+- auto-vectorization: Rely directly on the compiler vectorizer. In the case of [gcc](https://www.gnu.org/software/gcc/projects/tree-ssa/vectorization.html), it is typically enabled by the flag `-ftree-vectorize` and by default at `-O3`.
+
+Manual vectorization allows explicit control, but usually becomes architecture specific and assumes vector-like programming from the programmer.
+
+Auto-vectorization aims to improve the programmer's productivity by performing a compiler pass to automatically generate vector instructions in a given program. However, in practice help from the programmer is needed to achieve competitive vectorized codes, either fully manual or with the user annotations.
+
+An additional benefit of auto-vectorization and most user annotations is decoupling the program specification and underlying execution which is architecture-specific. The vectorizable source will be portable to different architectures and the compiler will dictate if it can be vectorized or how well it can exploit the program's DLP in order to generate the architecture-specific vector instructions.
+
+This project is aiming at having these general vector specifications at the compiler-level which:
+- permits intrinsics or vector assembly instructions to be translated into Bril IR
+- permits architecture specific backend to easily generate an executable
+- naturally expose opportunities for vector optimizations at Bril
+- offer to the compiler (automatic) vectorizer the possibility of generating Bril IR vector operations that can later be mapped into ISA vector instructions
+
+### Link to project
+[Vril](https://github.com/sa2257/vril) is our public repository.
+
+Arrays
+------------------
+
+### Arrays in Vril
+Vril includes a new type `array` to the language. The definition of an array in Vril is a sequence of elements of any type, which includes arrays (no type checking carried out) and the length being any Bril literal. 
+
+### Initializing arrays in Vril
+Vril extends Bril to support arrays by adding an `init` operator to initialize an array of length `l` and set its elements to 0 as follows:
+```
+array_name : array = init l
+```
+
+Vril adds array operations to Bril of the form,
+```
+variable: type = aop args
+```
+### Supported array operations
+Vril can operate on arrays in two ways: 1) We can perform *scalar* accesses to arrays by moving one element of the array to a variable and performing any arithmetic or logic operation that are already supported in the original Bril, and 2) we can perform *vector* accesses to the arrays by using new vector operations that take arrays as arguments directly.
+
+In order to move data from an element of an array `arr[idx]` to a variable `var`, we can use `a2v` op:
+```
+var: int = a2v array_name idx
+```
+Similarly, we can move a value stored in a variable `var` into one of the array's element `arr[idx]`:
+```
+array_name: int = v2a var index
+```
```

Comment:
No need to change this now, but it's a little weird that the array name appears in the "destination" position, i.e., on the left-hand side of the =. In every operation, the destination is where the result of the computation goes. But here the operation doesn't define a variable called `array_name`—instead, it puts a value into the array.

Reply:
You could see as an insert operation (there use to be one in earlier versions of RISC-V V): `vinsert v0, a1, a2` which does exactly that: `v0[$a1] = $a2`. Do you mean that it looks too much like assembly and less like an IR language?

Reply:
I see. Yes, it does not have a destination register notion here. The syntax you've suggested indeed makes that evident.